Motion Input Shaping with Streams

jayleong-zaber opened the pull request:
Added ability to perform input shaping with streams to execute more complex trajectories. A ShapedLockstep has also been created to allow for input shaping to be done on a lockstep axis. 

Both this new input shaping method and the pre-existing method based on timing of deceleration are still available. In order to manage which mode to use and different settings for each a ShaperConfig class was created.

Information for how to use the code and configure the shaper is provided in the README and more information on different shaper types is provided in input_shaper_types.md

Commit:
Add class to support performing input shaping on lockstep groups

Commit:
Fixed error in input shaping ShapedLockstep class

Commit:
Add more complex input shaping methods that exceuted using streams

Commit:
Fixed error with accel unit conversion and cleaned up formatting in ShapedAxis

Commit:
Moved disabling stream to start of move relative in ShapedAxis so that wait for idle works

Commit:
Replaced 2d arrays for time and accel pairs with individual variables in input shaping shtream generator for better clarity.
Edited and added comments to code.

Commit:
Add to example code in shaped_axis and shaped_lockstep to demonstrate using streams for shaping

Commit:
Update pipfile for motion_input_shaping

Commit:
Add details for stream input shaping to README

Commit:
Add document explaining the different type of input shapers and update images

Commit:
Add link to streams documentation to input shaping README

Commit:
Further changes to input shaping README

Commit:
Add shaper config to the settings of shaping_comparison_test.py to make it more clear that is something that can be changed

Commit:
Changes to shaper mode explanation in input shaping readme

Commit:
Reset com port number in shaping_comparison_test.py

Commit:
Run black code formatter and make changes suggested by pylint and mypy

Commit:
Changed initialization of impulses

Commit:
Simplify ShaperConfig settings dictionary to remove ability to access settings like they are attributes

Commit:
Remove min_timestep property that is unused and make impulses_updated private

Commit:
Create a specific shaper attribute for each mode to avoid a single attribute which different types possible. Add a shaper property to retrieve the relevant shaper if necessary

Commit:
Use enumeration in for loops in ShapedLockstep

Commit:
Refactor zero_vibration_stream_generator.py

Commit:
Changes to input_shaper_type.md

Commit:
Fixed errors flagged by pydocstyle

Commit:
Fixed mypy errors from matplotlib

Commit:
Add some additional details for using shaping_comparison_test.py to README

Commit:
Fixed some wording in input_shaper_types.md

andrewbrucelau-zaber committed:
Update Pipfile.lock to match versions of ZML for various platforms.

andrewbrucelau-zaber commented on src/motion_input_shaping/shaper_config.py:
```diff
@@ -0,0 +1,84 @@
+        :param options: Settings specified as keyword pairs
+        """
+        self.shaper_mode = shaper_mode
+        self._write_settings(options)  # process key word arguments and
```
and?

jayleong-zaber replied:
This class has been removed so this is no longer a problem

andrewbrucelau-zaber commented on src/motion_input_shaping/shaping_comparison_test.py:
```diff
@@ -9,13 +9,14 @@
+from matplotlib import pyplot as plt
 from zaber_motion import Units
 from zaber_motion.ascii import Connection
 from shaped_axis import ShapedAxis  # Zaber axis with input shaping
```
Comment probably unncessary

jayleong-zaber replied:
I have removed all the comments for these imports. It is all described in the docstrings anyways

andrewbrucelau-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,361 @@
+
+def trapezoidal_motion_generator(
+    distance: float, acceleration: float, deceleration: float, max_speed_limit: float
+) -> list[list[float]]:
```
t's not clear what is being returned (list[list[float]] is a bit ambiguous).  Consider defining namedtuple or dataclass?

jayleong-zaber replied:
I have added namedtuple (`AccelPoint`) to address this. I have changed some the other bits work with this and make everything more readable as well.

andrewbrucelau-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,361 @@
+    impulses: list[float],
+    unshaped_time: list[float],
+    unshaped_acceleration: list[float],
+) -> list[list[float]]:
```
not clear what is returned.  Consider namedtuple or dataclass?

jayleong-zaber replied:
Fixed with `AccelPoint` named tuple

andrewbrucelau-zaber reviewed (commented):
In file `zero_vibration_shaper.py`, there is setter and getter for `n`, which manipulates `self._n`, but everywhere else it just uses `self.n`.  Maybe it should be `self._n` instead?

grahamkerr-zaber commented on src/motion_input_shaping/README.md:
```diff
@@ -31,6 +31,7 @@ Alternatively, Python can be installed manually. This code is designed to run wi
 The following Python packages are used in this project:
 
 - Zaber Motion Library. Detailed installation information is available [here](https://software.zaber.com/motion-library/docs/tutorials/install/py).
+- Numpy. Detailed installation information is available [here](https://pypi.org/project/numpy/).
 - Matplotlib. Matplotlib is used for some of the examples involving plotting data. Detailed installation information is available [here](https://pypi.org/project/matplotlib/).
```
I'm on board with Bruce's recommendation. This was probably a bad idea in the first place on my part!

grahamkerr-zaber commented on src/motion_input_shaping/input_shaper_types.md:
```diff
@@ -0,0 +1,76 @@
+
+| Impulse Time | Impulse Magnitude |
+|--------------|-------------------|
+| $0$          | $\frac\{1}{1+K}$  |
+| $0.5T$       | $\frac\{K}{1+K}$  |
```
It looks fine on github!

grahamkerr-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -9,9 +9,11 @@
+from shaper_config import ShaperConfig, ShaperMode
 
 
 class ShapedAxis(Axis):
```
Followup from Martin's discussion this morning. We should change this to use composition instead of inheritance

jayleong-zaber replied:
Done

grahamkerr-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -46,6 +66,15 @@ def __init__(self, zaber_axis: Axis, resonant_frequency: float, damping_ratio: f
         self.reset_max_speed_limit()
 
+    @property
+    def shaper(self) -> ZeroVibrationShaper | ZeroVibrationStreamGenerator:
```
I'd be curious what someone more experienced thinks about this. Returning multiple object types seems like a sketchy practice, but it might be ok in python? Typically this is where you'd use an abstract base class or protocol to define a generic object.

jayleong-zaber replied:
I have eliminated this by splitting the class up in the `ShapedAxis` and `ShapedAxisStream`

grahamkerr-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -151,6 +215,86 @@ def move_relative_shaped(
             Units.ACCELERATION_MILLIMETRES_PER_SECOND_SQUARED,
         )
 
+    def _move_relative_shaped_stream(
```
I was about to add this same comment. We should just combine the two functions into move_relative_shaped()

jayleong-zaber replied:
These methods have now been moved into different classes

grahamkerr-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -151,6 +215,86 @@ def move_relative_shaped(
+        self.stream.setup_live(self.axis_number)
+        self.stream.cork()
+        for segment in stream_segments:
+            # Set acceleration making sure it is greater than zero by comparing 1 native accel unit
```
I think he's trying to make sure that accel could never round down to 0 in Zaber units if it was very low (0.01m/s/s). I do think this approach is most robust, but it also might be excessive? Never hurts I guess.

jayleong-zaber replied:
I did this for robustness and to prevent rounding down to 0 as Graham suggested. There are scenarios where this can happen if the accel is set very low to begin with or if the shaper is configured in a way that results in a segment of very low acceleration.

Commit:
Edit input shaping readme
- Add backticks
- Removed detail instructions for installing dependencies since all details are already precent in the pipfile
- Add link to zaber launcher
- Other minor changes

grahamkerr-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -222,6 +366,10 @@ def move_min_shaped(
 
 # Example code for using the class.
```
IMO some simple example code is nice, but Bruce does raise a good point. The vast majority of the code from this file is copied/pasted in shaped_lockstep.py. I think there should be a more elegant way to combine it. I have a feeling once we switch to composition instead of inheritance it will become more obvious. This may also be a weakness of ZML that a lockstep group can't be treated similar to a single axis (even though many of the methods are identical).

andrewbrucelau-zaber replied:
I agree that simple examples right in the class files are nice.

For this particular case, after the refactoring, how different is this piece of test code between `shaped_axis.py` and `shaped_axis_stream.py`?  If they are really similar, does it make sense to split them out in a separate file named `<something>_test.py` like the other files that we want users to run?

jayleong-zaber replied:
I have removed the example code from these files and created `shaping_demo.py` which does a simple demo of both classes and the methods in them.

I have also done a few changes to `shaping_comparison_test.py` so that is performs and plots shaping with both shaping classes so that they can be compared easily. In the plotting I added a subplot (the upper plot) to show the full trajectory of without zooming in so that it is easy to see the trajectory and the delay from communication and the device response on bottom that is zoomed in to see residual vibrations. An example is shown below but the device I used doesn't have a large oscillation so the response is not obvious.

![image](https://github.com/zabertech/zaber-examples/assets/144398940/52c34809-8ea7-4431-accc-35af91dc9816)

grahamkerr-zaber commented on src/motion_input_shaping/shaped_lockstep.py:
```diff
@@ -0,0 +1,549 @@
+"""
```
I think Jay is doing this specifically because he wanted to include a gantry example in a future article. I agree there is a ton of duplicated code though and it would be nice if we could combine it. That being said, having this file doesn't hurt. "Performs the same thing except for lockstep" is pretty intuitive.

jayleong-zaber replied:
This is now combined with `shaped_axis.py`

grahamkerr-zaber commented on src/motion_input_shaping/shaped_lockstep.py:
```diff
@@ -0,0 +1,549 @@
+            for n, axis in enumerate(self.axes):
+                axis.settings.set(setting, values[0], unit)
+
+    def get_axis_positions(self, unit: Units = Units.NATIVE) -> list[float]:
```
Do we need this? This functionality is essentially built into ZML. https://software.zaber.com/motion-library/api/py/ascii/lockstep#_getposition. I'm assuming you've done it for redundancy but I don't think it's necessary.

jayleong-zaber replied:
I use this to check the position of each axis in the group and then compare to the limit.max or min for each axis to determine the largest move possible (in the move_max_shaped and move_min_shaped methods). The lockstep get position in ZML gets the position of the primary axis but that might not be the limiting axis.

grahamkerr-zaber commented on src/motion_input_shaping/shaper_config.py:
```diff
@@ -0,0 +1,84 @@
+"""This file contains the ShaperConfig class which is used to specify shaping mode and settings."""
```
This code is pretty complex, but it's not intended for the user to modify, just use. Might be worth including that in the comments somewhere.

jayleong-zaber replied:
shaper_config has been completely removed in the latest changes

grahamkerr-zaber commented on src/motion_input_shaping/shaping_comparison_test.py:
```diff
@@ -26,6 +27,7 @@
 SCOPE_TIMEBASE = 1  # The scope sampling period in ms. Adjust if the whole move isn't captured.
 RESONANT_FREQUENCY = 5.07  # Input shaping resonant frequency in Hz.
 DAMPING_RATIO = 0.1  # Input shaping damping ratio.
+SHAPER_CONFIG = ShaperConfig(ShaperMode.DECEL)
```
Might want to add a comment here. We want users to know how to use this class without having to look at the source code.

jayleong-zaber replied:
Shaper config has been removed so this is no longer necessary

grahamkerr-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,361 @@
+        ]
+
+    # Trapezoidal Motion
+    max_speed = max_speed_limit
```
This variable is unused. Also, isn't most of the following calculation redundant with the one above? Can't you just set max_speed and max_speed_duration to the normal values, then re-calculate them (duration = 0) if max speed is not reached? saves a lot of lines.

jayleong-zaber replied:
I have made these changes and cleaned it up a bit.

grahamkerr-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,361 @@
+        # Initialize impulses as a single impulse with no delay for no shaping
+        self._impulse_times = [0.0]
+        self._impulses = [1.0]
+        self._impulses_updated = False
```
Is this needed? Why don't you just update the impulses automatically every time the frequency or damping ratio gets set? The calculation doesn't look expensive and those parameters won't be updated much. There's a ton of if/else logic relating to this variable that could be removed too.

jayleong-zaber replied:
I have opted to calculate them in get_impulses() and get_impulse_times(). You are right that it is not much computation so should be fine to recalculate every time. I'm not sure why I felt the need to do that in the first place.

grahamkerr-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,361 @@
+
+        match self.shaper_type:
+            case ShaperType.ZV:
+                a0 = 1 / (1 + k)
```
Write all these to the list directly instead of creating intermediate variables? Maybe I'm just being picky :)

jayleong-zaber replied:
Done

grahamkerr-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,361 @@
+        self._shaper_type = value
+        self._impulses_updated = False
+
+    def get_impulses(self) -> list[float]:
```
This method name is not the same as the non-stream shaper. It should be for consistency. Change one or the other.

grahamkerr-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,361 @@
+                raise ValueError(f"Shaper type {self.shaper_type} is not valid.")
+        self._impulses_updated = True
+
+    def shape_trapezoidal_motion(
```
A ton of these function definitions are extremely close to zero_vibration_shaper.py. I feel like there's so much redundant code. I wonder if n and calculate_n could be private in the other class? There's a chance we might be able to create fully identical class definitions for both classes, and maybe even simplify them to one class? 

Maybe it's not worth it. I just hate duplicating so much identical code.

jayleong-zaber replied:
There are a lot of elements shared between them but what is being done is also very different. For example, the impulse calculations look very similar but none of it can be shared between the two. I thought that keeping them separate is less confusing in a code example as it can become difficult to figure out which bits are actually relevant for a particular mode. I also see the argument for combining it all though.

One option to reduce the identical code could be to make a parent class for the stuff that is common between the two ( `resonant_frequency`, `damping_ratio`, and `resonant_period`).

jayleong-zaber replied:
`resonant_frequency`, `damping_ratio`, and `resonant_period` has been moved to a Plant class. Impulse calculations are still kept separate as they are not cross compatible.

grahamkerr-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,361 @@
+            self._update_shaper_impulses()
+        return self._impulse_times
+
+    def _update_shaper_impulses(self) -> None:
```
Actually... couldn't this just be split up so the calculation occurs in get_impulses() and get_impulse_times()?

jayleong-zaber replied:
Done

Commit:
Refactor src/motion_input_shaping/zero_vibration_stream_generator.py

Commit:
Run black code formatter on zero_vibration_stream_generator.py

jayleong-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -151,6 +215,86 @@ def move_relative_shaped(
+        stream_segments = self._shaper_stream.shape_trapezoidal_motion(
+            position_mm,
+            accel_mm,
+            accel_mm,
```
This needs to take decelonly setting

jayleong-zaber replied:
Done

jayleong-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,347 @@
+        self._shaper_type = shaper_type
+
+    @property
+    def resonant_frequency(self) -> float:
```
Create `Plant` class for these properties to be shared between shapers

jayleong-zaber replied:
Done

jayleong-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -46,6 +66,15 @@ def __init__(self, zaber_axis: Axis, resonant_frequency: float, damping_ratio: f
+                return self._shaper_stream
+
     @property
     def resonant_frequency(self) -> float:
```
Remove these properties that edit the shaper

jayleong-zaber replied:
Done

jayleong-zaber reviewed (commented):
Summary of plan from meeting:

- [x] Combine `ShapedAxis` and ShapedLockstep` using composition. 
     -  `ShapedAxis.move_relative()` for shaped moves and `ShapedAxis.axis.move_relative()` for unshaped moves
     - Type check using isinstance where necessary
- [x] Split into `ShapedAxis` and `ShapedAxisStream` 
     - Keeps the two as separate building blocks
    - This would eliminate the need for the config class

These changes have all been done

Commit:
Clean up code

jayleong-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -37,7 +45,19 @@ def __init__(self, zaber_axis: Axis, resonant_frequency: float, damping_ratio: f
             raise TypeError("Invalid Axis class was used to initialized ShapedAxis.")
```
Is this check necessary? `zaber_axis` is an instance of the `Axis` class so is it safe to assume that it is valid since it was successfully initialized before being passed into this?

jayleong-zaber replied:
It is possible to initialize `Axis` and `Lockstep` with an index that higher than available so this check is necessary

Commit:
Changed ShapedAxis use composition instead of inheriting Axis

Commit:
Use type of shaper in ShapedAxis to determine whihc mode is being used.

Commit:
Combine ShapedLockstep into ShapedAxis and split DECEL and STREAM modes into separate classes

Commit:
Use motion.decel in stream input shaping

Commit:
Add Plant class to define systems for input shaping

Commit:
Changes to input shaping readme

Commit:
Minor clean ups

Commit:
Removed type checking of shaper property since only a single type of shaper is being used in a class

jayleong-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -9,9 +9,11 @@
 
```
Done

jayleong-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -21,7 +23,13 @@ class ShapedAxis(Axis):
+        zaber_axis: Axis,
+        resonant_frequency: float,
+        damping_ratio: float,
+        shaper_config: ShaperConfig,
```
shaper_config is now no longer a param and I have updated the doc string with other changes

jayleong-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -246,14 +394,14 @@ def move_min_shaped(
 
         # Perform some unshaped Moves
```
These comments have been removed

jayleong-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -246,14 +394,14 @@ def move_min_shaped(
         time.sleep(1)
 
+        print("Shaping through changing deceleration.")
         # Perform some shaped Moves
```
These comments have been removed

jayleong-zaber commented on src/motion_input_shaping/shaped_axis.py:
```diff
@@ -280,4 +428,36 @@ def move_min_shaped(
+            ShaperConfig(ShaperMode.STREAM, shaper_type=ShaperType.ZV),
+        )  # Re-initialize ShapedAxis class using streams to perform shaping and specify ZV shaper
+
+        # Perform some shaped Moves
```
These comments have been removed

Commit:
Remove unnecessary comments

andrewbrucelau-zaber commented on src/motion_input_shaping/README.md:
```diff
@@ -52,51 +46,105 @@ To run the example code:
+- The `plant` parameter is an instance of the `Plant` class which defines the vibration that the input shaper is targeting.
+  - The `Plant` class has `resonant_frequency` and `damping_ratio` properties that define the target vibration frequency in Hz and damping ratio at which the input shaping algorithm will remove vibration.
+
+Class Methods:
```
All the `move_absolute_shaped()` methods have been changed to `move_absolute()`, i.e. minus `_shaped()`.  Need to update the documentation for both the ShapedAxis Class and ShapedAxisStream class.

jayleong-zaber replied:
Done

andrewbrucelau-zaber commented on src/motion_input_shaping/README.md:
```diff
@@ -52,51 +46,105 @@ To run the example code:
-- `move_relative_shaped()` - Moves to a relative position using a trajectory shaped for the target resonant frequency and damping ratio. Similar format to the [Axis.move_relative()](https://software.zaber.com/motion-library/api/matlab/ascii/axis#_MoveRelative) command.
-- `move_max_shaped()` - Moves to the max limit using a trajectory input shaped for the target resonant frequency and damping ratio. Similar format to the [Axis.move_max()](https://software.zaber.com/motion-library/api/matlab/ascii/axis#_MoveMax) command.
-- `move_min_shaped()` - Moves to the min limit using a trajectory input shaped for the target resonant frequency and damping ratio. Similar format to the [Axis.move_min()](https://software.zaber.com/motion-library/api/matlab/ascii/axis#_MoveMin) command.
+- `move_absolute_shaped()` - Moves to an absolute position using a trajectory shaped for the target resonant frequency and damping ratio. Similar format to the [Axis.move_absolute()](https://software.zaber.com/motion-library/api/py/ascii/axis#_moveabsolute) command.
```
Remove `_shaped()`

jayleong-zaber replied:
done

andrewbrucelau-zaber commented on src/motion_input_shaping/zero_vibration_shaper.py:
```diff
@@ -9,57 +9,21 @@
-        self.resonant_frequency = resonant_frequency
-        self.damping_ratio = damping_ratio
+        self.plant = plant
         self.n = 1  # How many periods to wait before starting deceleration.
```
The getter and setter of `self.n` manipulates `self._n`, but all the other functions utilizes `self.n`.  Should all the other functions use `self._n` instead?

jayleong-zaber replied:
I have made this change. I also got rid of the property setter since there is no need to modify `n` outside of the class and the error checking is not necessary with how it is computed.

andrewbrucelau-zaber commented on src/motion_input_shaping/zero_vibration_stream_generator.py:
```diff
@@ -0,0 +1,311 @@
+    ZVDD = 3
+
+
+@dataclass
```
Just curious, why did you choose to use `@dataclass` for `StreamSegment`, but `NamedTuple` for `AccelPoint`?

jayleong-zaber replied:
In line 287, I set the final position of the last segment to the target position in order to remove any small rounding errors. I used `dataclass` for `StreamSegment` to make that easier since `NamedTuple` is immutable. I don't think there is a reason for `AccelPoint` to specifically be one or the other so I'll change that to `dataclass` for consistency.

Commit:
Change AccelPoint to dataclass for consistency

Commit:
Removed property setter for n since it never needs to be modified from outside of the class

Commit:
Removed "_shaped" from method names in README

Commit:
Corrected typo in readme

Commit:
Removed demo code from shaped_axis and shaped_axis_stream and created shaping_demo.py that demos both.

Commit:
Modified shaping_comparison_test.py to test both shapers and added subplot to display full trajectory

Commit:
Remove unused imports

Commit:
Replaced "_test" suffix in filenames with "_demo"

andrewbrucelau-zaber reviewed (approved):
Looks good to me!

Commit:
Fixed reference to wrong class names in error messages
